### DISCUSSION TITLE
CancellationTokenSource.CreateLinkedTokenSource does not have an overload with TimeProvider

### DISCUSSION THREAD

**Discussion by ramonsmits** [dotnet/runtime#95822] [Q&A] [answered]
I often have code that does something like:

```c#
async Task MyAmazingLogic(CancellationToken cancellationToken)
{
	using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
	timeoutCts.CancelAfter(TimeSpan.FromSeconds(10));

	// Invoke API's
	await InvokeOtherStuff(timeoutCts.Token);
}
```

I wanted to use this together with `TimeProvider` but `CreateLinkedTokenSource` does not have an overload.

It seems I can only do:

```c#
TimeProvider timeProvider;

async Task MyAmazingLogic(CancellationToken cancellationToken)
{
    using var timeoutCts = new CancellationTokenSource(Timeout, timeProvider);
	using var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
	
	// Invoke API's
	await InvokeOtherStuff(combinedCts.Token);
}
```

It would be nice if `CreateLinkedTokenSource` would support `TimeProvider`. Even better if you could just do the following as in a lot of cases I just want to use it to create a timeout that also adheres to cooperative cancellation.

```c#
TimeProvider timeProvider;

async Task MyAmazingLogic(CancellationToken cancellationToken)
{
	using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, TimeSpan.FromSeconds(10), timeProvider);
	
	// Invoke API's
	await InvokeOtherStuff(timeoutCts.Token);
}
```

Did I miss any API's? Is the "double" CTS creation the only way to do this with .net 8?

**Comment by stephentoub** [comment, marked as answer]
>  Is the "double" CTS creation the only way to do this with .net 8?

Or just do the registration yourself, since that's really all CreateLinkedTokenSource does for you (including disposing of the registration when you dispose the CTS):
```C#
async Task MyAmazingLogic(CancellationToken cancellationToken)
{
    using var cts = new CancellationTokenSource(Timeout, timeProvider);
    using var _ = cancelationToken.Register(s => ((CancellationTokenSource)s).Cancel(), cts);

    // Invoke API's
    await InvokeOtherStuff(cts.Token);
}
```

**Reply by ramonsmits** [reply]
Is there any gain on using `Register` over `CreateLinkedTokenSource`? It looks just a little more complex although it pretty much the same behavior.

ps. I was you used `using var _` and was hoping that would be a discard variable but unfortunately its not. Still pretty clear though when not using the value.

**Reply by stephentoub** [reply]
If you use UnsafeRegister instead of Register, it's the same thing.
